### ISSUE TITLE
Eliminate non-generic functions

### ISSUE THREAD

**Issue by masaeedu** [Microsoft/TypeScript#17428]
Every non-generic function can be expressed as generic function with a single type parameter corresponding to each argument. Explicit parameter type annotations simply correspond to `extends` constraints on the type parameters of the generic equivalent.

While the title of the issue is "eliminate non-generic functions", in practice this simply gets rid of all the syntax ceremony involved in declaring generic functions. A function declaration as follows:

```
function repeat(item, n: number) => Array(n).fill(item)
```

will be inferred as: `type F = <T1 extends any, T2 extends number>(item: T1, number: T2) => T1[]`, and the result of invoking it with `repeat("foo", 10)` is inferred as `string[]`, not `any[]`.

**Comment by ahejlsberg** [comment]
> Every non-generic function can be expressed as generic function with a single type parameter corresponding to each argument.

That is not correct. For example, in a function taking two parameters of type `Foo`, those two parameters have *identical* types and are assignable to each other. But if you substitute two type parameters (each constrained to `Foo`), neither is assignable to the other because they may actually represent distinct and different types.

**Comment by masaeedu** [comment]
@ahejlsberg Given that TypeScript doesn't have ref returns or out variables, I don't really see where it would be useful to specifically assign one variable to another. Within the function body, both `a` and `b` are assignable to references of type `Foo`. Additionally, since the proposal is to eliminate non-generic functions, and not generic functions, it is still possible to express `<T1 extends Foo>(x: T1, y: T1)`.

**Comment by gcnew** [comment]
Is this an ask for non-local type inference in disguise? I don't see why already annotated parameters should be synthetically made polymorphic.

**Comment by masaeedu** [comment]
@gcnew It is mainly a syntactic request rather than a semantic one. I want to hijack straightforward function definition syntax for declaring polymorphic functions, because in 99% of cases the resulting polymorphic function types are just as good (if not better) for people intending to write monomorphic functions. The current syntax for writing polymorphic functions like `function repeat(item, n: number) => Array(n).fill(item)`, which uses `<>`, is extremely unwieldy and nests poorly.

I'm not specifically asking for some new HM-esque inference strategy, although that would be easier to set up if you're already abstracting over and generating constraints for each parameter type. In this issue I'm just seeking a reduction in ceremony for declaring generic functions.

**Comment by gcnew** [comment]
I don't see how polymorphic types could be assigned to parameters without doing the actual inference. Or do you mean only in the cases when the type parameters need no constraints? Also, I don't completely agree with your views on `<>`, it's an explicit `forall` which doesn't take more space than in other languages. On the contrary, it adds clarity. On numerous occasions people have said "AHA" when I've translated signatures to the `<>` style.

**Comment by masaeedu** [comment]
> Unfortunately, I don't see how polymorphic types could be assigned to parameters without doing the actual inference

The type arguments are simply inferred from the types of the value arguments passed at the callsite, just as they are in today's generic functions. Nothing about the semantics of generic functions changes, we just start inferring generic function types for function expressions that were previously ascribed monomorphic function types. 

E.g. the expression `const id = a => a;` is ascribed the type `<T>(a: T) => T` instead of the type `(a: any) => any`, and so at a callsite my application of `id(10)` results in proper inference of `number` for `T`.

> Or do you mean only in the cases when the type parameters need no constraints?

Perhaps I'm missing something, but the reasoning above seems to work even for cases where we have (subtype) constraints. After all, having no constraints is just the special case of being constrained to the top type (`{}`?). E.g. if I write the function `const f = (a: A) => Object.assign({}, a, { foo: a.aStuff() })`, passing in a `B extends A` at the callsite would result in the more useful result `B & { foo: ... }`.

> Also, I don't completely agree ...

The explicit `forall` can add clarity in some cases, so it is good to still have the ability to use `<>`, but there is a reason Haskell has the lowercase letter shorthand for declaring type parameters. If you had to explicitly write a `forall` quantified type definition for every tiny function you'd never get anything done.

I'm working with React higher-order components right now, and beyond a couple levels of nesting the type definitions basically flow off the screen. And this is *with* total separation of the type and implementation a la `type Foo = ...; const foo: Foo = ... /* no explicit types here */`.

**Comment by gcnew** [comment]
You might find https://github.com/Microsoft/TypeScript/issues/15114 interesting. TypeScript's unconstrained unions (i.e. unions between arbitrary types, not a predefined set of `data` constructors), intersections, overloads, mutability, subtyping, literals, etc. makes type reconstruction quite challenging. I'm not saying it's completely impossible, but it would be extremely hard to do right with the current state of TS. Especially without a proper unification solver.

Explicit `forall` is not so uncommon. E.g. PureScript has made it mandatory.

I've never done anything remotely serious with React, but I'd expect HKTs to be a bigger pain point than long parametric definitions. My guess is that parameter names are also big part of the problem (https://github.com/Microsoft/TypeScript/issues/13152). For the latter, I've been recently thinking about "shorthand" type annotations (inspired by Haskell):

```ts
const id :: a -> a -> a = x => x; // `::` denotes a shorthand type annotation
type Id = :: a -> a -> a
```

Not quite sure the new syntax would be worth the complexity and segregation it's adding.

**Comment by masaeedu** [comment]
Is the `::` a special syntax that activates "shorthand" type definitions? I'd personally be happy with that too, I'm just finding life very difficult when I define curried functions of several arguments and I need to keep performing the magic `<T extends Foo>(foo: T)` ritual to prevent losing type information at each step. The point of parameter type annotations (for me at least) is to constrain the types of input, not to throw them away.

I'll look at #15114, although I don't see how that is relevant to this issue. Could you make this a bit more concrete? I understand that TypeScript generating unions out of thin air makes things difficult with respect to normalizing types, but that doesn't totally erase the utility of parametrically polymorphic functions; there's still many cases where doing `<T extends Foo>(foo: T)` retains better type fidelity than `(foo: Foo)`.

**Comment by gcnew** [comment]
> Is the `::` a special syntax that activates "shorthand" type definitions?

Yes, that was my intention.

>The point of parameter type annotations (for me at least) is to constrain the types of input, not to throw them away.

Unfortunately TypeScript has some difficulties with parametric polymorphism, but on the upside things have improved significantly.

> I'll look at #15114, although I don't see how that is relevant to this issue.

To be honest, I think this suggestion is a dupe of #15114. However, even if it weren't, the discussion there provides insight what makes deducing polymorphic types hard, e.g. https://github.com/Microsoft/TypeScript/issues/15114#issuecomment-293393165.

**Comment by masaeedu** [comment]
@gcnew It isn't a dupe of #15114, because that is asking for changes to type inference, whereas I'm asking for changes to syntax. To put this another way, if you were to implement what is being asked for in this issue, you'd only touch `parser.ts`; `checker.ts` would remain totally unchanged.

Whether or not deducing polymorphic types is hard is irrelevant; we have a simple mechanical way of transforming non-generic function declarations into generic function declarations at the AST level. How well or poorly type inference works with generic functions once they've been declared is a discussion for a different issue.

**Comment by gcnew** [comment]
I've understood your suggestion wrong. You've used the word _inferred_ several times which had brought me a wrong impression. Now I see that your actual suggestion is to assign fresh type parameters to every function parameter that doesn't have an explicit type provided, instead of the current implicit `any`. The parameters that already have a (non polymorphic) type would be converted to bounds.

```ts
function f(a, b, c: number) {
    return a + c;
}
```

Would be treated as if it were

```ts
function f<A, B, C extends number>(a: A, b: B, c: C) {
    return a + c; // an error here according to the existing rules
}
```

I have mixed feelings on usability. Without inferring constraints (actual inference based on usage) on the introduced type parameters, I'm doubtful it would be very useful. And it doesn't seem very backward compatible either. On the other hand, it's a step in a more strict and sound direction.

**Comment by masaeedu** [comment]
@gcnew Yes, that's exactly it; sorry about the confusion. I realized halfway through the conversation that "inferred" was the wrong word, perhaps "ascribe" is better.

The intention is to make this as backward compatible as possible. As @ahejlsberg has pointed out, it isn't quite backward compatible if you intend to assign to parameters. However if you treat parameters strictly as input positions (i.e. you treat a function as a *contravariant* generic type), this should be mostly backwards compatible. Wherever a parameter of type monomorphic type `X` is being *consumed*, a parameter ascribed the polymorphic type `T extends X` will do just as well.

Additionally, this would "standardize" parameter typing in a way that would set you up for improved inference, as proposed in #15114. E.g. when type checking a function body, for every parameter with no explicit bounds, you could simply emit additional bounds on the generic type where you would previously have emitted an error.

It is true that within a function expression of the form `function f(x: Foo)`, it now becomes impossible to assign to `x`. However, it should be noted that in JS, assigning to `x` isn't really particularly useful, unless you feel like being stingy with local variables. JS has no concept of "pass by reference" or "out params", so reassignment of parameters is guaranteed to have no impact on the caller's scope. For this reason I think requiring anyone reassigning their parameters to use `as any` is an acceptable tradeoff.

If this modest break with backwards compatibility is deemed unacceptable, we have two alternatives that maintain total backwards compatibility:
 - Only apply this treatment to parameters with no explicit type annotations, which is basically #14078, and comes with some awkward tradeoffs
 - Improve the expressiveness of bounds on parametric polymorphism, so that it is possible to express both `function<T extends Foo>` and `function<Foo extends T>`, which, as a corollary, gives you `function<T is Foo>`. Then this proposal, except with `T is AnnotatedType` instead of `T extends AnnotatedType`.

**Comment by kitsonk** [comment]
> On the other hand, it's a step in a more strict and sound direction.

Actually I would argue that `noImplicitAny` is a better way, and good news is we already have it. 😁

**Comment by masaeedu** [comment]
@kitsonk The purpose is to make declaring polymorphic functions less verbose. E.g. this:

```
const pluck = <TProp extends string>(p: TProp) => <T extends {[k in TProp]}>(x: T) => x[p]
```

Would now be expressed as:

```
const pluck = (p: string) => (x: {[k in typeof p]}) => x[p]
```

I don't see how `noImplicitAny` helps you with this.

**Comment by simonbuchan** [comment]
Perhaps a compromise?

```
function repeat(item extends any, n extends number) => Array(n).fill(item)
const pluck = (p extends string) => (x extends { [k in typeof p] }) => x[p];
```

**Comment by mhegazy** [comment]
As [noted by @ahejlsberg](https://github.com/Microsoft/TypeScript/issues/17428#issuecomment-318113602), the proposal does not take into consideration the difference in meaning between generic types parameters and other non-generic types. under this proposal, two strings are not comparable any more, since:
```ts
function NonGeneric(a: string, b: string) {
    a == b; // OK
}

function Generic<T extends string, U extends string>(a: T, b: U) {
    a == b; // Not OK, since T and U are not comparable
}
```
And that seems to be pretty fundamental.

Moreover, as [noted by @gcnew](https://github.com/Microsoft/TypeScript/issues/17428#issuecomment-318176595) languages that use [Hindly-milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system) type systems get a lot of mileage of this idea because the combine it with inference from call sites, and unifying constraints. just getting the everything-is-generic part puts a lot of constraints on  function implementors, and makes using these types fairly onerous. 

Having a new syntax to define generic type parameters does not seem to be the right solution either. adding new constructs/syntax to the language increases learning and maintenance costs for both users and compiler maintainers.

**Comment by masaeedu** [comment]
@mhegazy It seems like a bug for `T` and `U` to not be comparable when both are subtypes of `string` (which does support `==`), but I see your point.

**Comment by mhegazy** [comment]
If they are generic, they are two distinct types, and not the same one. e.g. call `Generic<"foo", "bar">` these are not comparable.

also see some related discussions in https://github.com/Microsoft/TypeScript/issues/17926 and https://github.com/Microsoft/TypeScript/issues/11218

**Comment by masaeedu** [comment]
@mhegazy That's well and good, but they are both still instances of at least `string`, and should be comparable as such. Right now I can do `(a as string) === b`, but this explicit cast is redundant; the compiler is already aware that both `a` and `b` are `string`.

**Comment by masaeedu** [comment]
Putting it another way, if types are sets, `a` and `b` are values selected from two unknown *subsets* of `string`. Thus any operations that demand strings (such as `===(a: string, b: string): boolean`) should be satisfied with `T`s and `U`s.

**Comment by tycho01** [comment]
@masaeedu: yeah, it does feel like the `Generic` example is a more fundamental problem about these comparisons, rather than about the use of generics here. e.g. `const one = 1; declare let num: number; one == num` should be legitimate in my view, rather than raising a compiler error unless `num` is guaranteed to be within `1`.

It does feel hard to get right though -- I get that they wanted to error on `num as number == str as string`. The question feels like at what level to compare things. e.g. in your `a == b` example, the compiler would need to know whether to compare the types as they are (distinct generics), widened (`string`), or something even higher (like `any`, at which point the check becomes moot again). Say `[1,2,3]` could be widened to a few different things, though for comparator cases like this just `number[]` should likely do.

I do wonder where else this might matter. I think I can help with that question, as I've tried something similar in my PR for unwidened `const` (https://github.com/Microsoft/TypeScript/pull/17785#issuecomment-322973594). As noted here, that mostly seems to affect functions like `<T>(a: T, b: T)`. When for that PR I tried the change on TS's test suite, this meant I had to annotate a dozen places like that throughout, essentially the test `*.ts` files at the bottom of https://github.com/tycho01/TypeScript/commit/30c9beb6f012c5990b215deefe81700b8c7e3599.

I think an opt-in proposal can be an improvement even with cons: if this proposal could improve inference for many scenarios, I believe having to annotate a few other cases would be fair, given a new compiler flag so users could opt in at their own leisure.
The concerns raised are definitely relevant, but similar reasoning could have been used to disregard `strictNullChecks`.

**Comment by masaeedu** [comment]
@tycho01 I don't know if operators are given special treatment in TypeScript, but I'd expect it to work exactly the same as though I had:

```
interface Equatable<T> {
    equals(other: Equatable<T>): boolean
}

class A implements Equatable<A>
{
    equals(o) {
        // ...
    }
}

class B extends A { }
class C extends A { }

declare const b: B;
declare const c: C;

b.equals(c); // No type error
```

We could say "there is no distinct `B#equals(c: C)` operation, but this doesn't matter, because TypeScript is able to determine some set of supertypes from the operands for which the operation is defined. Similar logic is needed for `===` and other operators.

**Comment by masaeedu** [comment]
Actually, looks like that example doesn't quite work how I expect it to; I was assuming `o` would be inferred to `Eq<A>`, but apparently not. Here's a simplified example:

```

**Comment by tycho01** [comment]
@masaeedu:

> TypeScript is able to determine some set of supertypes from the operands for which the operation is defined.

It grabs an `equals` definition with `A` (not `B`) for `T`, hence `c` matches.

> I don't know if operators are given special treatment in TypeScript

For `==` see `checkBinaryLikeExpression` in `checker.ts` (too big to view on Github) -- for `EqualsEqualsToken` it checks `isTypeEqualityComparableTo` in both directions for limited-widened (`getBaseTypeOfLiteralType`) versions of the operand types.

> Similar logic is needed for === and other operators.

The idea sounds interesting. I tried it by throwing a `getApparentType` on top, which makes `Generic` pass, see https://github.com/tycho01/TypeScript/commit/36d39fb118f16e1759d094edb8f72a871c6c3889.
Other affected tests (https://github.com/tycho01/TypeScript/commit/7eb7f499cacf0df5e9cdacd4510715b1a54008b5):
- `0 == 1` now allowed
- `T == U` now allowed unless their constraints indicate we should believe otherwise
- existing errors like `Operator '==' cannot be applied to types 'number' and 'string'` somehow switched to capitalized types like `Operator '==' cannot be applied to types 'Number' and 'String'`. Not sure I get this one.

Overall I'm not seeing any real damage with this particular change, so hopefully that could help the original proposal here back on track.

**Comment by gcnew** [comment]
The conversation has diverted towards https://github.com/Microsoft/TypeScript/issues/17445.

@tycho01 
> existing errors like Operator '==' cannot be applied to types 'number' and 'string' somehow switched to capitalized types like Operator '==' cannot be applied to types 'Number' and 'String'. Not sure I get this one.

That's because of your use of `getApparentType`.

**Comment by tycho01** [comment]
I feel like the rejections have held some circular logic; the current thread was closed since #17445 was yet unresolved, then that thread got closed for apparent lack of remaining use-cases...

**Comment by RyanCavanaugh** [comment]
Reopening

**Comment by Raiondesu** [comment]
A compromise suggested by @simonbuchan seems to me like a great backward-compatible way to implement this, even though I do see the point of the syntax/semantics change that @masaeedu proposes.

<details>
<summary>The case that @mhegazy mentioned</summary>

is really actually a bug or at least an unexpected behaviour and has no relevance to this issue:
```ts

function equal<T extends string, U extends string>(a: T, b: U) {
  return a == b;
  // TS says "This condition will always return 'false' since the types 'T' and 'U' have no overlap"
}

// Here, however, the overlap is clearly present:
equal('foo', 'foo'); // inferred as `equal<'foo', 'foo'>(a: 'foo', b: 'foo'): boolean`
// 'foo' and 'foo' overlap perfectly well,
// So what gives? C'mon TS, I know you're better than this! 😏
```

Even proves true for a much less edgy case:
```ts
// I see code like this in production and libraries all the time:
const enum FooBar {
  Foo = 'foo',
  Bar = 'bar'
}

declare const foobar: FooBar;

equal(foobar, FooBar.Foo);
// Yup, still overlap perfectly and type-check correctly.
// Hence, the error in the implementation of `equal` must be incorrect, I guess?
```

Maybe create a separate issue for this?
</details>

Solving this issue is very relevant to me and could also probably/partially solve https://github.com/microsoft/TypeScript/issues/14400.

So, here I go, coming up with my take on two proposals for this:

## # 1 - Implicitly-generic function parameters

<details>
<summary>This is, basically, what @masaeedu suggests, if I understand the issue correctly.</summary>

This proposal, however, needs evaluation in terms of backward-compatibility, as I'm not sure that it is actually backward-compatible.

### Functions

All functions in TS are generic by default and infer their parameters implicitly.

A function definition like
```ts
declare function add(x: number, y: number): number;
```
is the same as writing
```ts
declare function add<X extends number, Y extends number>(x: X, y: Y): number;
```
in current TS.

Writing "classic" generics explicitly simply allows for a finer control of the function declaration by extracting repeating types from the definition into type variables.\
For example, in this scenario:
```ts
declare function addObjectKeys(
  obj1: { [key: string]: number },
  obj2: { [key: string]: number },
  key: keyof typeof obj1 & keyof typeof obj2
): number;
```
the declaration sure does get very tedious to read.\
So we need a way to extract the repeated type into a "type variable" of sorts, just like how we would do in current TS:
```ts
declare function addObjectKeys<O extends { [key: string]: number }>(
  obj1: O,
  obj2: O,
  key: keyof O
): number;
```
which would be equivalent to writing this in current TypeScript:
```ts
declare function addObjectKeys<
  O extends { [key: string]: number },
  Obj1 extends O,
  Obj2 extends O,
  Key extends keyof O
>(
  obj1: Obj1,
  obj2: Obj2,
  key: Key
): number;
```

This ensures that all functions are treated the same way, and current generic syntax becomes basically a place for defining type variables for functions.

Most generics in type-heavy code (properly typed code) are used for this exact purpose - declaring type variables for later use in conditionals and stuff.

So, I'd say that actually the semantics of generics do change.\
But I can only see positive effects in this.

### Classes
Classes and functions in JavaScript and TypeScript are used nearly interchangeably by many developers, which makes the possible scope of this issue wider than just functions.

<details>
<summary>So, if it <b>does</b> touch classes, then this section becomes relevant.</summary>
 
Using this syntax in classes is a big concern, as it looks like it would require a big rework of how the class generics are currently handled.

With this syntax in mind, constructors now must have generics in order to satisfy the monomorphic conversion:
```ts
// This:
class Box<T> {
  constructor(public param: T) {}
}

// Should now mean this:
class Box<T> {
  constructor<Param extends T>(param: T) {}
}
```

Such a constructor would not affect the instantiation of the class, and use its generic parameters only as type variables:
```ts
new Box<'foo' | 'bar'>('foo');
// Generic `constructor` has no effect on the final constructor function invocation
```

Inferring constructor parameters would work just as it does now:
```ts
// This:
class Box<T> {
  constructor(public param: T) {}
}
new Box('foo'); // Box<string>

// Should now mean this:
class Box<T> {
  constructor<Param extends T>(param: T) {}
}
new Box('foo'); // Box<string>
```
 
</details>

</details>

## # 2 - Explicitly-generic function parameters with syntax sugar

<details>
<summary>This is, basically, what @simonbuchan suggests.</summary>

Contrary to the previous one, this proposal doesn't change the semantics of generics, but rather the syntax of parameter's type definitions.

All functions in TS stay the same, nothing changes, full backward-compatibility.

However, the `extends` keyword is now allowed after any function parameter and `infer` is allowed after `extends`.\
So, function definitions like
```ts
declare function add(x extends number, y extends number): number;
```
are just a syntax sugar for
```ts
declare function add<x extends number, y extends number>(x: x, y: y): number;
```
The short equivalent produces types with the same names as the parameters they correspond to.

This proposal allows for very flexible and easy-to-read generic function definitions:
```ts
type Foo = { foo: number };

// this beauty:
declare function addToFoo(foo extends Foo, x: number): number;

// instead of this pile of boilerplate code:
declare function addToFoo<foo extends Foo>(foo: foo, x: number): number;
```

Another addition of such syntax would be that the parameter can now "be referenced" as types, because the desugared type names are the exact same as the parameter names:
```ts
declare function map<T>(array extends Array<T>, f extends (value: T) => any): Array<ReturnType<f>>;
// equivalent to:
declare function map<T, f extends (value: T) => any>(array extends Array<T>, f: f): Array<ReturnType<f>>;
```

The addition of `infer` into the mix is not necessary, but makes everything just so gorgeous:
```ts
// even better so:
declare function map(array extends infer T[], f extends (value: T) => infer R): R[];
// equivalent to this mouthful:
declare function map<
 T,
 f extends (value: T) => any,
 R extends f extends (value: T) => infer R ? R : never
>(array: T[], f: f): R[];
```

So, the `infer` keyword in parameters is a shorthand for conditional types in the same way that `extends` is a shorthand for generic types.

However, old typescript definitions can't take advantage of this, as **no new semantics** for existing syntax are introduced.\
But, as a nice bonus, this proposal brings zero breaking changes!

### Classes
Classes and functions in JavaScript and TypeScript are used nearly interchangeably by many developers, which makes the possible scope of this issue wider than just functions.

<details>
<summary>So, if it <b>does</b> touch classes, then this section becomes relevant.</summary>

`extends` in constructor parameters is disallowed just like generics, but are **still allowed in methods** and properties.

```ts
// This should produce a compiler error!
class Box<T> {
  constructor(
				// here
	public param extends T
  ) {}
}

// Because it's just syntax sugar for
class Box<T> {
  constructor<param extends T>(public param: param) {}
}
// which is illegal
```

```ts
class Box<T> {
  constructor(public value: T) {}

  map(f extends (value: T) => infer R): Box<R>;
}
// equivalent to:
class Box<T> {
  constructor(public value: T) {}

  map<f extends (value: T) => any, R extends f extends (value: T) => infer R ? R : never>(f: f): Box<R>;
}
```

So, no breaking changes here either.

</details>

My personal favourite is this one, as it brings no breaking changes, while also allowing developers to write generic functions much more easily, greatly increasing the probability that a developer would prefer to write a nice generic function:
```ts
declare function someFunction(param extends any);
```
instead of this:
```ts
declare function someFunction(param: any);
```
as practically no extra code is introduced in the process,\
making this proposal's syntax **greatly** superior to simple `param: type` definitions.

This would also allow linters to be configured in a way to favor the new syntax over the "old" one, increasing overall code quality in projects.
</details>

---

I tried to represent as many syntactic variations as I could while also keeping the amount of text reasonably small.

I'm not very strong in writing EBNF definitions, so none are present, please, pardon me here.\
Any questions/additions are welcome, and I'm willing to continue on completing these proposals' definitions.
